File: iloadins/test_instrument_alt_output_note.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

// chan: 1..20
// board_pos: 1..29
static bool test_instrument_alt_output_note (uint8_t chan, uint8_t board_pos, uint8_t ins1, uint8_t ins2)
{
  bool result = false;
  int_least16_t note;
  uint16_t freq;

  DBG_ENTER ("test_instrument_alt_output_note");

  chan--; // chan: 0..19
  board_pos--; // board_pos: 0..28

  note = board_pos + 12 * (current_octave - 1);
  if ((note < 0) || (note > 12*8+1)) goto _exit;

  chan_handle[chan] = board_scancodes[board_pos];
  if (ins2 != 0) chan = _4op_main_chan[chan] - 1;
  load_instrument (&temp_instrument, chan + 1);

  if (   percussion_mode
      && (   (temp_instrument.perc_voice == 4)
          || (temp_instrument.perc_voice == 5)))
    load_instrument (&temp_instrument, _perc_sim_chan[chan - 18]);

  if (ins2 != 0)
    load_instrument (&temp_instrument2, chan - 1);

  macro_speedup = songdata.macro_speedup;
  if (play_status == isStopped) update_timer (songdata.tempo);

  freq = nFreq (note - 1) + 0x2000 + ((int8_t *)&temp_instrument)[12];
  event_table[chan].note = note;
  freq_table[chan] = freq;
  freqtable2[chan] = freq;
  key_on (chan + 1);
  change_freq (chan + 1, freq);

  init_macro_table (chan + 1, note, ins1, freq);
  if (ins2 != 0)
  {
    freq_table[chan - 1] = freq;
    freqtable2[chan - 1] = freq;
    init_macro_table (chan - 1 + 1, note, ins2, freq);
  }

  result = true;

_exit:
  DBG_LEAVE (); //EXIT //test_instrument_alt_output_note
  return result;
}

File: adt2ext3/pas/test_instrument_alt_output_note.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

// Local to `test_instrument_alt' procedure.

function output_note(chan,board_pos: Byte): Boolean;

var
  note: Byte;
  freq: Word;

begin
  _dbg_enter ({$I %FILE%}, 'test_instrument_alt.output_note');

  note := board_pos+12*(current_octave-1);
  If NOT (note in [0..12*8+1]) then
    begin
      output_note := FALSE;
      _dbg_leave; EXIT; //test_instrument_alt.output_note
    end;

  chan_handle[chan] := board_scancodes[board_pos];
  If (test_ins2 <> 0) then chan := _4op_main_chan[chan];
  load_instrument(temp_instrument,chan);

  If percussion_mode and
     (temp_instrument.perc_voice in [4,5]) then
    load_instrument(temp_instrument,_perc_sim_chan[chan]);

  If (test_ins2 <> 0) then
    load_instrument(temp_instrument2,PRED(chan));

  macro_speedup := songdata.macro_speedup;
  If (play_status = isStopped) then update_timer(songdata.tempo);

  freq := nFreq(note-1)+$2000+
          SHORTINT(pBYTE(@Addr(temp_instrument)^)[12]);
  event_table[chan].note := note;
  freq_table[chan] := freq;
  freqtable2[chan] := freq;
  key_on(chan);
  change_freq(chan,freq);

  If (test_ins2 = 0) then init_macro_table(chan,note,test_ins1,freq)
  else begin
         init_macro_table(chan,note,test_ins1,freq);
         freq_table[PRED(chan)] := freq;
         freqtable2[PRED(chan)] := freq;
         init_macro_table(PRED(chan),note,test_ins2,freq);
       end;

  output_note := true;

  _dbg_leave; //EXIT //test_instrument_alt.output_note
end;
